Add .finally to dojo/Deferred

RoyTinker opened the pull request:
This implements ES2018 `finally` in dojo/Deferred.

Includes comprehensive tests in tests/unit/Deferred.js (all passing).

See #351 for discussion - this opts for option (1).

dylans commented:
Hi Roy, this approach seems reasonable to me. I'm going to leave the PR open for a bit in case anyone else has feedback they would like to give.

wkeese commented:
Looks good to me.  "finally" is certainly a useful feature, and you have good unit tests.

I've never seen that "void 0" thing before.   Looked it up just now.  Seems like you are being overly-paranoid that someone has redefined `undefined` to mean something else.   (And if they have, the rest of dojo won't work anyway.)

RoyTinker commented:
@wkeese Thanks for reviewing. I replaced `void 0` with `undefined` in the unit tests, and fixed a couple of issues I found with my added documentation.

RoyTinker commented on Deferred.js:
```diff
@@ -262,6 +262,63 @@ define([
+					var valueOrPromise = callback();
+					if (valueOrPromise && typeof valueOrPromise.then === "function"){
+						return valueOrPromise.then(function (){
+							return new Deferred().reject(reason);
```
@dylans @wkeese An alternative here (also on line 307) is to use `throw reason`. There may be a slight performance difference, but that may vary between platforms.

Any preference?

wkeese replied:
I don't know enough about the code to comment...

RoyTinker replied:
@dasa Anyone from your team want to take a look?

dasa replied:
I'd probably go with `throw reason;` since I don't see any other code in this module using `new Deferred().reject` but there is code that throws in strict mode.

RoyTinker replied:
Thank @dasa. I switched to `throw reason`

RoyTinker commented:
2 updates:
- fixed some test issues - now all are passing (I wasn't looking in the right place previously)
- switched to `throw reason` from `return new Deferred().reject(reason)` on lines 304 and 307

@dylans @wkeese I think this is ready to merge.

RoyTinker commented:
@dylans Could we get this merged?

RoyTinker commented:
Closing in favor of https://github.com/dojo/dojo/pull/362 -- see that PR description for reasoning.
